File: src/main/java/easy/Q0009.java
package easy;

/**
 * 9. Palindrome Number
 * https://leetcode.com/problems/palindrome-number/
 */
class Q0009 {

  // O(n) time, O(1) space, where n is the number of digits in `x`
  public boolean isPalindrome(int x) {
    if (x < 0) {
      return false;
    }
    String xStr = Integer.toString(x);
    for (int i = 0; i < Math.floorDiv(xStr.length(), 2); i++) {
      if (xStr.charAt(i) != xStr.charAt(xStr.length() - 1 - i)) {
        return false;
      }
    }
    return true;
  }
}

File: src/main/rust/easy/Q0009.rs
/// 9. Palindrome Number
/// https://leetcode.com/problems/palindrome-number/


/// Time: O(n), Space: O(n), where n is the number of digits in `x`.
pub fn is_palindrome(x: i32) -> bool {
    let x = x.to_string().into_bytes();
    let (mut lo, mut hi) = (0, x.len() - 1);
    while lo < hi {
        if x[lo] != x[hi] {
            return false;
        }
        lo += 1;
        hi -= 1;
    }
    true
}
